### ISSUE TITLE
Docs improvements for improps

### ISSUE THREAD

**Issue by petsuter** [imageio/imageio#956]
`improps()` seems (in theory) like a very cool addition to the v3 API. 🚀  Improving the docs on this would be great:

* [improps](https://imageio.readthedocs.io/en/stable/_autosummary/imageio.v3.improps.html) should link to [ImageProperties](https://imageio.readthedocs.io/en/stable/_autosummary/imageio.core.v3_plugin_api.ImageProperties.html) and vice versa.
* The [examples](https://imageio.readthedocs.io/en/stable/examples.html) should have an example with `improps`.
* Most important: `spacing` should document the type (`Tuple[float,...]`?) and unit (millimeters?)!

Are there any plugins that actually provide spacing yet? I only found tifffile_v3 and that seems to return just a value but not the unit, which can be different from file to file even in tiff format. That is obviously rather useless unfortunately. 😢

**Comment by FirefoxMetzger** [comment]
> - [improps](https://imageio.readthedocs.io/en/stable/_autosummary/imageio.v3.improps.html) should link to [ImageProperties](https://imageio.readthedocs.io/en/stable/_autosummary/imageio.core.v3_plugin_api.ImageProperties.html) and vice versa.
> - The [examples](https://imageio.readthedocs.io/en/stable/examples.html) should have an example with improps.

I agree. I've added and example to https://github.com/imageio/imageio/pull/959 as well as a reference to ImageProperties in the docs of `improps`.

> Most important: spacing should document the type (Tuple[float,...]?) and unit (millimeters?)!

`ImageProperties.spacing` exists because I keep getting feedback that this number should be exposed, but I never used it myself (not a biologist), so I am a bad judge of what a good format for this is. I was hoping that someone with either experience or a need for it would start chiming in to tell me what's missing and then we tweak ImageProperties to make `spacing` useful, but that did not happen until now.

For now, it is exposing a tuple of type `Tuple[Union[Tuple[float, ...], float], ...]`. If the spacing is constant for a particular axis then the spacing at that index is a `float`, otherwise, it is a `Tuple[float, ...]` indicating the spacing between elements.

@petsuter If the unit can vary then we probably also need a `ImageProperties.spacing_units`. Any suggestions of how a good format for these should look like?

**Comment by petsuter** [comment]
> example to #959 as well as a reference to ImageProperties in the docs of `improps`.
Nice. 👍 

> a tuple of type `Tuple[Union[Tuple[float, ...], float], ...]`. If the spacing is constant for a particular axis then the spacing at that index is a `float`, otherwise, it is a `Tuple[float, ...]` indicating the spacing between elements.

Do you have examples of non-constant spacing? It sounds very exotic, and I would leave this to the format-specific metadata in `immeta` and leave the standardized `ImageProperties.spacing` as `None` for simplicity.

> If the unit can vary then we probably also need a `ImageProperties.spacing_units`. Any suggestions of how a good format for these should look like?

For my use case an explicit `spacing_mm: Tuple[float, ...]` would be most useful where a a fixed unit (millimeter) is used, and the plugins have to convert to that fixed unit or use `None` if converting is not possible. This would also be simple to implement and has precedents:

* DICOM Pixel Spacing Attribute is in millimeter.
* SimpleITK GetSpacing assumes millimeter.
* nibabel get_zooms() uses millimeter.
* TIFF Tag ResolutionUnit allows inch or centimeter.
* Various other specialized formats I know use a fixed millimeter, micrometer or nanometer unit.

In general there could be some downsides to using a fixed unit like millimeter, e.g. not every image has axis units that can be converted to a metric distance (time, frequency, ...), and for very small (nanometer resolution microscopy images) or large (kilometer resolution geography imagery) values the float numbers would use up ~6 of the ~15 significant digits in a 64 bit float just for zeros. It would Probably not be suitable for astronomy images, see `astropy.units`.

A `spacing_units` can make sense, e.g. a `Tuple[str, ...]` with a set of predefined unit strings like `"mm"` could be used. But what other units to predefine? This gets quite complex quickly and should then probably be coordinated with e.g. https://discuss.scientific-python.org/t/physical-units-summit/390

Instead of separate attributes for numeric value and unit, astropy and others define quantity types so `Tuple[Quantity, ...]` can keep both numeric values and units in the same attribute.

**Comment by FirefoxMetzger** [comment]
> Do you have examples of non-constant spacing? It sounds very exotic, and I would leave this to the format-specific metadata in immeta and leave the standardized ImageProperties.spacing as None for simplicity.

If, for example, one creates a TIFF from an image time series with non-uniformly spaced time points. One might apply a treatment to a sample, take pictures after 0.5h, 1h, 2h, 6h, 12h, 24h, and store the result in a TIFF. This person may wish to load and operate these images as a batch of `(time, height, width, channel)` images and if `spacing` is needed here, the temporal spacing would be non-constant.

> For my use case an explicit spacing_mm: Tuple[float, ...] would be most useful where a a fixed unit (millimeter) is used

This segues into using a constant spacing everywhere (e.g. `mm`). The main issue is that we have to (at least) worry about 3 types of axes: spatial, temporal, and color. For the axis that represents color, I think it would be okay to just return `None` and revisit once somebody has an actual use for this. For the temporal axis, seconds seem like a natural choice; however, nanoseconds are a good alternative that would match the default timestamp unit. For spatial axes, it is less clear to me if this should be meter, inch, mm, or nm.

Another idea could be to add `ImageFormat.spacing_scale`, which would hold the conversion factor to convert it into a SI unit. For `mm` spacing this would then be `1e3` and users that want `mm` and can't make the assumption that `spacing` will already be in `mm` can use `props.spacing * props.spacing_scale * 1e-3`.

I am musing a bit, but this might also be a good time to investigate if we can reliably extract the type of axis and report something like `props.axes = ("spatial", "spatial", "color")`. This is useful in general to programmatically discover what kind of image one is dealing with (planar/packed, 3D volume, time-series, ...) and I think it would help make reasoning about `spacing` easier, too.

**Comment by petsuter** [comment]
Yes, differentiating color channels from spatial axes is useful. (In OME-TIFF they call this DimensionOrder e.g. XYZTC and the "scaling" values are e.g. PhysicalSizeX, PhysicalSizeY, PhysicalSizeZ, TimeIncrement, WaveStart, and WaveIncrement. WaveStart hints at more complexity due to the "absolute vs. delta" question I think. They define >30 length units alone, >20 time units, and a LOT more.)

Agree, for (color) channels a spacing makes no sense I think. (Again there's a lot more complexity in e.g. colorspace, calibration and physical units for what the pixel intensity values mean.)
